packages/appium/lib/extension/manifest.js

```diff
@@ -0,0 +1,570 @@
+ * @returns {value is ExtensionPackageJson<PluginType>}
+ */
+function isPlugin (value) {
+  return isExtension(value) && _.isString(_.get(value, 'appium.pluginName'));
+}
+
+/**
+ * Handles reading & writing of extension config files.
+ *
+ * Only one instance of this class exists per value of `APPIUM_HOME`.
+ */
+export class Manifest {
+  /**
+   * The entire contents of a parsed YAML extension config file.
+   *
+   * Contains proxies for automatic persistence on disk
+   * @type {ManifestData}
+   * @private
+   */
+  _data;
+
+  /**
+   * Path to `APPIUM_HOME`.
+   * @private
+   * @type {Readonly<string>}
+   */
+  _appiumHome;
+
+  /**
+   * Path to `extensions.yaml`
+   * @type {string}
+   * Not set until {@link Manifest.read} is called.
+   */
+  _manifestPath;
+
+  /**
+   * Helps avoid writing multiple times.
+   *
+   * If this is `undefined`, calling {@link Manifest.write} will cause it to be
+   * set to a `Promise`. When the call to `write()` is complete, the `Promise`
+   * will resolve and then this value will be set to `undefined`.  Concurrent calls
+   * made while this value is a `Promise` will return the `Promise` itself.
+   * @private
+   * @type {Promise<boolean>|undefined}
+   */
+  _writing;
+
+  /**
+   * Helps avoid reading multiple times.
+   *
+   * If this is `undefined`, calling {@link Manifest.read} will cause it to be
+   * set to a `Promise`. When the call to `read()` is complete, the `Promise`
+   * will resolve and then this value will be set to `undefined`.  Concurrent calls
+   * made while this value is a `Promise` will return the `Promise` itself.
+   * @private
+   * @type {Promise<void>|undefined}
+   */
+  _reading;
+
+  /**
+   * Sets internal data to a fresh clone of {@link INITIAL_MANIFEST_DATA}
+   *
+   * Use {@link Manifest.getInstance} instead.
+   * @param {string} appiumHome
+   * @private
+   */
+  constructor (appiumHome) {
+    this._appiumHome = appiumHome;
+    this._data = _.cloneDeep(INITIAL_MANIFEST_DATA);
+  }
+
+  /**
+   * Returns a new or existing {@link Manifest} instance, based on the value of `appiumHome`.
+   *
+   * Maintains one instance per value of `appiumHome`.
+   * @param {string} appiumHome - Path to `APPIUM_HOME`
+   * @returns {Manifest}
+   */
+  static getInstance = _.memoize(function _getInstance (
+    appiumHome,
+  ) {
+    return new Manifest(appiumHome);
+  });
+
+  /**
+   * Searches `APPIUM_HOME` for installed extensions and adds them to the manifest.
+   * @param {SyncWithInstalledExtensionsOpts} opts
+   * @returns {Promise<boolean>} `true` if any extensions were added, `false` otherwise.
+   */
+  async syncWithInstalledExtensions ({depthLimit = DEFAULT_SEARCH_DEPTH} = {}) {
+    const walkOpts = _.defaults({depthLimit}, DEFAULT_FIND_EXTENSIONS_OPTS);
+    // this could be parallelized, but we can't use fs.walk as an async iterator
+    let didChange = false;
+    for await (const {stats, path: filepath} of fs.walk(
```

Comment:
is this an interview question? 😄   this is my best guess; I've never seen the second syntax, but it looks valid.

```js
for await (x of y) { // y is an async iterator object. y is _not_ treated like a `Promise`
  // the promise is here. the loop does not continue until it resolves
}
```

```js
for (x of await y) {  // y may or may not be a promise, but the loop doesn't begin until it resolves to a synchronous iterator.
  // no promise in here
}
```

Reply:
ok. i think you thought i was being smarter than i am. but your comment helped me to see the answer is "async iterators".

i've only ever used the second syntax. i guess i thought we needed it because it was a promise that resolved to an array. but instead it is apparently an async iterator. so we want the first syntax.